sdk/state/ingest.go

```diff
@@ -169,3 +169,116 @@ func (c *Channel) ingestTxMetaToUpdateBalances(resultMetaXDR string) error {
+	var initiatorEscrowAccountEntry, responderEscrowAccountEntry xdr.AccountEntry
+	var initiatorEscrowTrustlineEntry, responderEscrowTrustlineEntry xdr.TrustLineEntry
+	for _, o := range txMeta.V2.Operations {
+		for _, change := range o.Changes {
+			updated, ok := change.GetUpdated()
+			if !ok {
+				continue
+			}
+
+			// Grab any trustline updates in case the channel asset is non-native.
+			trustline, ok := updated.Data.GetTrustLine()
+			if ok {
+				if trustline.AccountId.Address() == c.initiatorEscrowAccount().Address.Address() {
+					initiatorEscrowTrustlineEntry = trustline
+				} else if trustline.AccountId.Address() == c.responderEscrowAccount().Address.Address() {
+					responderEscrowTrustlineEntry = trustline
+				}
+			}
+
+			account, ok := updated.Data.GetAccount()
+			if !ok {
+				continue
+			}
+			if account.AccountId.Address() == c.initiatorEscrowAccount().Address.Address() {
+				initiatorEscrowAccountEntry = account
+			} else if account.AccountId.Address() == c.responderEscrowAccount().Address.Address() {
+				responderEscrowAccountEntry = account
+			}
+		}
+	}
+
+	// If initiator escrow account not found, return.
+	var empty xdr.AccountId
+	if initiatorEscrowAccountEntry.AccountId == empty {
+		return nil
+	}
+
+	// Validate the initiator escrow account sequence number is correct.
+	if int64(initiatorEscrowAccountEntry.SeqNum) != c.startingSequence {
+		return fmt.Errorf("incorrect initiator escrow account sequence number found, found: %d want: %d",
+			int64(initiatorEscrowAccountEntry.SeqNum), c.startingSequence)
+	}
+
+	escrowAccounts := [2]xdr.AccountEntry{initiatorEscrowAccountEntry, responderEscrowAccountEntry}
+	for _, ea := range escrowAccounts {
+		// Validate the escrow account thresholds are correct.
+		if ea.Thresholds != xdr.Thresholds([4]byte{0, requiredNumOfSigners, requiredNumOfSigners, requiredNumOfSigners}) {
+			return fmt.Errorf("incorrect initiator escrow account thresholds found")
+		}
+
+		// Validate the escrow account has the correct signers and signer weights.
+		var initiatorSignerFound, responderSignerFound bool
+		for _, signer := range ea.Signers {
+			address, err := signer.Key.GetAddress()
+			if err != nil {
+				return fmt.Errorf("parsing formation transaction escrow account signer keys: %w", err)
+			}
+
+			if address == c.initiatorSigner().Address() && signer.Weight == xdr.Uint32(1) {
+				initiatorSignerFound = true
+			} else if address == c.responderSigner().Address() && signer.Weight == xdr.Uint32(1) {
+				responderSignerFound = true
+			}
+		}
+		if !initiatorSignerFound || !responderSignerFound {
+			return fmt.Errorf("incorrect signer weights found")
+		}
+		if len(ea.Signers) != requiredNumOfSigners {
+			return fmt.Errorf("incorrect number of signers, found: %d want: %d", len(ea.Signers), requiredNumOfSigners)
+		}
+	}
+
+	// Validate the trustlines are correct.
+	// TODO - should an initiator escrow be allowed to have extraneous trustlines (if not add test)?
```

Comment:
@leighmcculloch wdyt about a native asset channel EI having existing trustlines? I can't think of an explicit reason to now allow it, but might not be thinking of something

Reply:
There's two routes we can go:
1. Verify that an account is in a very specific state, only has the trustline required, has specific flags, etc.
2. Or, verify that only the things we need are set the way we need.

I think we should do the latter because the first approach will always be a moving target. For example, when accounts get liquidity pools, do we check for those too? I don't think we want that sort of burden if we can avoid it.

I think we should do our best to ignore anything that has no bearing on the security of the channel. This means:
- Verifying that there are only the signers we specify, this is one thing we must check full state on.
- Verifying that the trust line for the channel's asset is has an appropriate balance that other than auth revocation is definitely available to supply the closing transaction's payment. That includes taking into account trust line liabilities, so I've opened a new issue #251 to cover adding liabilities.
- Verifying that the trust line for the channel's asset is authorized. \*
- Ignoring everything else.

\* We could ignore the authorization status of the trust line if we hope to support [SEP-8](https://stellar.org/protocol/sep-8) in the manner described in https://github.com/stellar/experimental-payment-channels/discussions/34#discussioncomment-765861. We aren't implementing support for SEP-8 right now so I think it makes sense to check authorization in our current implementation even if the authorization can be revoked anytime, and we can remove it if we go down a path of implementing SEP-8.